### PR TITLE
feat: Implement gRPC server to ingest streaming features

### PR DESCRIPTION
**Pull request by mehmettokgoz**
**What this PR does / why we need it**:

Currently only way to ingest streaming features to Feast is using the Spark/Kafka processor. This processor has several limitations. It retrieves data in batches and forces to run all transformations in Python runtime. 

The gRPC server that listens for the streaming features allows users to apply transformation on more powerful streaming engines and send it to Feast then. This feature introduces new streaming sources for Feast, in addition to Kafka and Kinesis.

Our main purpose for the gRPC ingestion service is to integrate Hazelcast streaming engine with Feast. Hazelcast provides various [connectors](https://docs.hazelcast.com/hazelcast/5.3/integrate/connectors), and supports to process those sources in real-time. This server makes possible to sink transformed data to Feast.

### CODE REVIEW

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (adchia) [inline sdk/python/feast/infra/contrib/grpc_server.py:43]: the max workers probably should be made configurable?

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (adchia) [inline sdk/python/feast/infra/contrib/grpc_server.py:1]: any thoughts on combining this with the "feature server", which can optionally be reused here? is the main reason to separate it out because you want to make a grpc server?

Comment (adchia) [reply sdk/python/feast/infra/contrib/grpc_server.py:1]: i.e. https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L82

Comment (mehmettokgoz) [reply sdk/python/feast/infra/contrib/grpc_server.py:1]: Yes, we maintain gRPC connectors so HTTP is out of support.

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (mehmettokgoz) [reply sdk/python/feast/infra/contrib/grpc_server.py]: Makes sense: https://github.com/feast-dev/feast/pull/3687/commits/ba1d045d652cbb9483b262542cd5ccda1cf482d0

**Comment by mehmettokgoz** [comment]
Thanks for review @adchia. Waiting for your answers.

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (adchia) [inline sdk/python/feast/infra/contrib/grpc_server.py:39]: why limit to just one sfv? seems better to keep it flexible?

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (mehmettokgoz) [reply sdk/python/feast/infra/contrib/grpc_server.py]: removed. ✅

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (mehmettokgoz) [reply sdk/python/feast/infra/contrib/grpc_server.py]: ✅

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (mehmettokgoz) [reply sdk/python/feast/infra/contrib/grpc_server.py]: added heath check service to server. ✅

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (mehmettokgoz) [reply sdk/python/feast/infra/contrib/grpc_server.py]: makes sense. added feature_view name to proto. ✅

File: sdk/python/feast/infra/contrib/grpc_server.py
Comment (mehmettokgoz) [reply sdk/python/feast/infra/contrib/grpc_server.py:1]: refactored to have similar logic with regular feature server. ✅

**Comment by mehmettokgoz** [comment]
@adchia addressed review comments. ✅

Changes:
- Do not force only one feature view. Proto object contains a `feature_view_name` field.
- Push API is supported.
- Health check service is added to server.
- `grpcio` packages are upgraded to 1.56.2
- CLI is refactored.

**Comment by crispin-ki** [comment]
I saw a comment about tests failing in this PR in the community [notes](https://docs.google.com/document/d/1zUbIWFWjaBEVlToOdupnmKQwgAtFYx41sPoEEEdd2io/edit#heading=h.nwsa98vsvv4u). 

Failing python unit tests should be fixed by https://github.com/feast-dev/feast/pull/3734. Once that PR is merged you can rebase this branch on top of master and the failing tests should pass.

**Comment by mehmettokgoz** [comment]
Thanks @crispin-ki 

Do you have any idea why linter might be failing? The errors are not related to the changes in this PR and I see it's green for your PR.

**Comment by crispin-ki** [comment]
> Thanks @crispin-ki
> 
> Do you have any idea why linter might be failing? The errors are not related to the changes in this PR and I see it's green for your PR.

Yeah it's a bit weird that these are failing now, seeing as you didn't touch them. I think you have a few options though:
- logs for failing workflow (pasted at bottom of this comment too) are all E721s, so you could add `E721` [here](https://github.com/feast-dev/feast/blob/e81684d4f7916c986fa8e6cf06c2918951469799/setup.cfg#L13) so those errors are now ignored altogether
- alternatively you could go to each line that is failing and add `# noqa: E721` at the end of line similar to [here](https://github.com/Ki-Insurance/feast/blob/7af776aa73e16bf24e62e836ad3a418c652bb475/sdk/python/feast/on_demand_feature_view.py#L72C1-L72C1) but with E721 instead, so flake doesn't check that line for E721
- alternatively you could go to each line that is failing and change them so flake is happy (in most cases change `==` to `is`)

In terms of which is preferred, I think all have pros and cons. I think just go with whichever you feel is most appropriate (or if unsure, might be one for maintainers to decide)

Logs for the failing workflow:
```
cd /home/runner/work/feast/feast/sdk/python; python -m flake8 feast/ tests/
feast/diff/registry_diff.py:[13](https://github.com/feast-dev/feast/actions/runs/5887487203/job/15966956374?pr=3687#step:9:14)6:16: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
feast/feature_store.py:1109:12: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
feast/field.py:65:12: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
feast/infra/offline_stores/bigquery.py:728:12: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
feast/type_map.py:748:8: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
tests/unit/local_feast_tests/test_local_feature_store.py:472:20: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
tests/unit/test_feature_views.py:183:[16](https://github.com/feast-dev/feast/actions/runs/5887487203/job/15966956374?pr=3687#step:9:17): E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
```

**Comment by mehmettokgoz** [comment]
@crispin-ki  

I found that it is related to new version of flake8 (which upgrades pycodestyle). Now pycodetype check for EE721 (https://github.com/PyCQA/pycodestyle/blob/main/CHANGES.txt#L12). I set flake version to `>=6.0.0,<6.1.0`.

**Review by adchia — APPROVED** [review]

**Review by adchia — APPROVED** [review]

**Pull request merged**
